File: Java/CTCI/src/linked_list/LinkedList.java
package linked_list;
import java.util.*;
 

public class LinkedList {

	protected Node head = null;
	
	protected int length = 0;
	
	public LinkedList() {
		length = 0;
		head = null;
	}
	
	public boolean compare(LinkedList other) {
		if (length != other.length) {
			return false;
		}
		Node n1 = head;
		Node n2 = other.head;
		while(n1 != null && n2 != null) {
			if (n1.data != n2.data) {
				return false;
			}
			n1 = n1.next;
			n2 = n2.next;
		}
		return true;
	}
	public LinkedList(int len, boolean duplicate) {
		for (int i= 0; i < len ;i++) {
			this.prepend(i);
			if (duplicate)
				this.prepend(i);
		}
	}
	
	public LinkedList(int len) {
		for (int i= 0; i < len ;i++) {
			this.prepend(i);
		}
	}
	
	public boolean contains(int data) {
		if (head == null) {
			return false;
		}
		Node temp = head; 
		while(temp != null) {
			if (temp.data == data) {
				return true;
			}
			temp = temp.next;
		}
		return false;
	}

	public int get(int index) throws IndexOutOfBoundsException, IllegalArgumentException {
		if (index < 0) {
			throw new IllegalArgumentException("k cannot be less than or equal to zero");
		}
		
		if (head == null || index >= length) {
			throw new IndexOutOfBoundsException("List has " + length + " elements, cannot return " + index + "th element");
		}
		
		Node temp = head;
		int count = 0;
		while (count < index) {
			temp = temp.next;
			count++;
		}
		return temp.data;
	}
	
	public void prepend(int data) {
		Node n = new Node(data);
		if (head == null) {
			head = n;
		} else {
			n.next = head;
			this.head = n;
		}
		length++;
	}
	
	public int getLength() {
		return length;
	}
	
	public void print() {
		Node temp = head;
		while(temp != null) {
			System.out.print(temp.data + " ");
			temp = temp.next;
		}
		System.out.println();
	}
	
	// to print circular lists, give option to print first <len> nodes
	public void print(int len) {
		Node temp = head;
		int count = 0;
		while(temp != null && count < len) {
			System.out.print(temp.data + " ");
			temp = temp.next;
			count++;
		}
		System.out.println();
	}
	
	public void remove_dups() {
		if (head == null) {
			return;
		}
		HashSet<Integer> vals = new HashSet<Integer>();	
		Node prev = head;
		Node temp = head.next;
		vals.add(head.data);
		while(temp!= null) {
			if (vals.contains(temp.data)) {
				prev.next = temp.next;
				length--;
			} else {
				prev = temp;
				vals.add(temp.data);
			}
			temp = temp.next;
		}
	}
	public void remove_dups2() {
		if (head == null) {
			return;
		}
		
		Node valnode = head;
		while (valnode != null) {
			Node prev = valnode;
			Node temp = valnode.next;
			int val = valnode.data;
			while(temp!= null) {
				if (val == temp.data) {
					prev.next = temp.next;
					length--;
				} else {
					prev = temp;
				}
				temp = temp.next;
			}
			valnode = valnode.next;
		}
	}
	
	public Node kth_last_node(int k) throws IndexOutOfBoundsException, IllegalArgumentException {
		if (k <= 0) {
			throw new IllegalArgumentException("k cannot be less than or equal to zero");
		}
		
		if (head == null) {
			throw new IndexOutOfBoundsException("List is empty, cannot return " + k + "th element");
		}
		
		Node to_end = head;
		Node kth_last = head;
		int count = 0;
		
		while((to_end != null) && (count < k)) {
			to_end = to_end.next;
			count++;
		}
		
		if (count < k) {
			throw new IndexOutOfBoundsException("List is " + count + " elements long, cannot return " + k + "th element");
		}
		
		while (to_end != null) {
			to_end = to_end.next;
			kth_last = kth_last.next;		
		}
		return kth_last;
	}
	
	public void partition(int k) {
		boolean passed = false;
		Node temp = head;
		Node prev = head;
		while(temp != null) {
			if (temp.data < k) {
				
				if (passed) {
					prev.next = temp.next;
					temp.next = head;
					head = temp;
					temp = prev.next;
				} else {
					prev = temp;
					temp = temp.next;
				}
			} else {
				passed = true;
				prev = temp;
				temp = temp.next;
			}
		}
	}
	public Node loop_detection() {
		Node n1 = head;
		Node n2 = head;
		boolean passed = false;
		while(n1 != null && n2 != null) {
			if (n1 == n2) {
				if (!passed) { // we're starting at head so they're going to collide at least once
					passed = true;
				} else {
					// loop detected
					n1 = head;
					while(n1 != n2) {
						n1 = n1.next;
						n2 = n2.next;
					}
					return n1;
				}
			}
			n1 = n1.next; // n1 increment by 1
			n2 = n2.next; // n2 increment by 2 
			if (n2 != null) { // watch out for loop ending here, unchecked by "while"
				n2 = n2.next; 
			} else {
				return null; // tail reached, no loop exists
			}
		}
		return null; // tail reached, no loop exists
	}
	
	// node cannot be head or last element
	public static void remove_middle_node(Node n) {
		Node temp = n.next;
		n.data = temp.data;
		n.next = temp.next;
	}
	
	public void reverse_in_place() {
		if (head == null)
			return;
		int[] stack = new int[length];
		Node temp = head;
		int count = 0;
		while(temp!= null) {
			stack[count++] = temp.data;
			temp = temp.next;
		}
		temp = head;
		count--;
		while(count >= 0) {
			temp.data = stack[count--];
			temp = temp.next;
		}
	}
	public static Node isIntersecting(LinkedList l1, LinkedList l2) {
		Node t1 = l1.head;
		Node t2 = l2.head;
		Node t1_prev = null;
		Node t2_prev = null;
		int len1 = 0;
		int len2 = 0;
		while (t1 != null) {
			t1_prev = t1;
			t1 = t1.next;
			len1++;
		}
			
		while(t2 != null) {
			t2_prev = t2;
			t2 = t2.next;
			len2++;
		}
		if (t1_prev != t2_prev ) {
			return null;
		}
		int len_diff = Math.abs(len1 - len2);
		Node n1;
		Node n2;
		if (len_diff != 0) {
			LinkedList longer = len1 > len2 ? l1 : l2;
			LinkedList shorter = len1 > len2? l2 : l1;
			int k = (len1 > len2 ? len1 : len2) - len_diff;
			n1 = longer.kth_last_node(k);
			if (n1 == null) {
				System.out.println("kth_last_node failed");
				return null;
			} else {
				n2 = shorter.head;
			}
		} else {
			n1 = l1.head;
			n2 = l2.head;
		}
		while(n1 != null && n2 != null) {
			if (n1 == n2) {
				return n1;
			}
			n1 = n1.next;
			n2 = n2.next;
		}
		return null;
	}
	public boolean isPalindrome() {
		if (head == null) {
			System.out.println("The list is empty");
			return false;
		}
		int[] stack = new int[length];
		Node temp = head;
		int count = 0;
		while (temp != null) {
			stack[count++] = temp.data;
			temp = temp.next;
		}
		temp = head;
		count--;
		while (count >= 0) { 
			if (temp.data != stack[count--]) {
				return false;
			}
			temp= temp.next;
		}
		return true;
	}
}

File: Python/02_Linked_List/LinkedList.py
class Node:
    def __init__(self, data):
        self.data = data;
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None
        self.length = 0

    def print(self):
        temp = self.head
        while(temp is not None):
            print(str(temp.data) + " ", end='')
            temp = temp.next
        print("\n")

    def prepend(self, data):
        if self.head == None:
            self.head = Node(data)
            self.length += 1
        else:
            temp = Node(data)
            temp.next = self.head
            self.head = temp
            self.length += 1

    def __init__(self, list_length):
        self.head = None
        self.length = 0
        for i in range(0, list_length):
            self.prepend(i)
            self.prepend(i)
        

    def remove_duplicates(self):
        if self.head is None:
            return

        vals = set()
        vals.add(self.head.data)
        #print(vals)
        prev = self.head
        temp = self.head.next

        while(temp is not None):
            if temp.data in vals:
                prev.next = temp.next
                self.length-= 1
            else:
                vals.add(temp.data)
                prev = temp

            temp = temp.next
    
    def remove_duplicates2(self):
        if self.head is None:
            return

        
        valnode = self.head
        while(valnode is not None):
            prev = valnode
            val = valnode.data
            #print(str(val))
            temp = prev.next
            while(temp is not None):
                if temp.data == val:
                    prev.next = temp.next
                    self.length-=1
                else:
                    prev = temp
                temp = temp.next
            valnode = valnode.next

    def kth_last_node(self, k):
        if self.head is None:
            print("List is empty, cannot return " + str(k) + "th element")
            return False, 0

        if k == 0:
            print("K cannot be 0")
            return False, 0

        to_end = self.head
        k_last = self.head
        count = 0
        while(to_end is not None and count < k):
            to_end = to_end.next
            count += 1

        if count < k:
            print("List is " + str(count) + " elements long, cannot return " + str(k) + "th element")
            return False, 0

        while to_end is not None:
            to_end = to_end.next
            k_last = k_last.next

        return True, k_last

    def partition(self, k):
        passed = False
        temp = self.head
        prev = self.head
        while(temp is not None):
            if (temp.data < k):
                if not passed:
                    prev = temp
                    temp = temp.next
                    
                else:
                    prev.next = temp.next
                    temp.next = self.head
                    self.head = temp
                    temp = prev.next
            else:
                passed = True
                prev = temp
                temp = temp.next
                


# the node given is not the first nor last node of the list
def remove_middle_node(n):
    temp = n.next
    n.data = temp.data
    n.next = temp.next



l1 = LinkedList(10)
l1.print()
l1.remove_duplicates2()
l1.print()
is_answer, node = l1.kth_last_node(5)
if is_answer:
    print("answer is " + str(node.data))
    remove_middle_node(node)
    l1.print()
    l1.partition(4)
    l1.print()